File: 08-1.rb
#!/usr/bin/env ruby

require 'ostruct'

lines = File.read('08.input').lines.map(&:strip)

line_format = /^(?<from>[[:alpha:]]+) = \((?<left>[[:alpha:]]+), (?<right>[[:alpha:]]+)\)$/

insns = lines[0].chars.map do |i|
  if i == 'L'
    'left'
  else
    'right'
  end
end

nodes = {}

lines[2..].each do |line|
  m = line.match line_format
  nodes[m['from']] = OpenStruct.new({ left: m['left'], right: m['right'] })
end

node = 'AAA'
ip = 0
steps = 0
until node == 'ZZZ'
  node = nodes[node][insns[ip]]
  ip = (ip + 1) % insns.size
  steps += 1
end

print steps, "\n"

File: 08-1.py
#!/usr/bin/env python

import re
from collections import namedtuple

lines = [row.strip() for row in open('08.input').readlines()]

line_format = re.compile("^(?P<from>[A-Z]+) = \\((?P<left>[A-Z]+), (?P<right>[A-Z]+)\\)$")

insns = lines[0]

Node = namedtuple('Node', ['L', 'R'])
nodes = {}
for line in lines[2:]:
    m = line_format.match(line)
    nodes[m['from']] = Node(L=m['left'], R=m['right'])

node = 'AAA'
ip = 0
steps = 0
while node != 'ZZZ':
    node = getattr(nodes[node], insns[ip])
    ip = (ip + 1) % len(insns)
    steps += 1

print(steps)
